hail/python/hailtop/aiotools/fs.py

```diff
@@ -634,6 +639,10 @@ async def copy_as_dir(self, sema: asyncio.Semaphore, source_report: SourceReport
             self.src_is_dir = False
             await self.release_barrier()
             return
+        except:
+            self.encountered_error = True
+            await self.release_barrier()
+            raise
```

Comment:
I think I'd write this like this:

```
    async def copy_as_dir(self, sema: asyncio.Semaphore, source_report: SourceReport, return_exceptions: bool):
        try:
	    src = self.src
	    if not src.endswith('/'):
		src = src + '/'
	    try:
		srcentries = await self.router_fs.listfiles(src, recursive=True)
	    except (NotADirectoryError, FileNotFoundError):
		self.src_is_dir = False
		return
	    self.src_is_dir = True
        finally:
            await self.release_barrier()

        await self.barrier.wait()

        if self.src_is_file is not None and self.src_is_file:
            raise FileAndDirectoryError(self.src)

        ...
```

This makes the release of the barrier clearer.  Then we can remove release_barrier_and_wait.  We need the `self.src_is_file is not None` in the case `copy_as_file` got an error.